File: packages/indexer/cognis_indexer/parsers/_normalize.py
"""AST text normalizer and content hash utilities.

The normalizer is the foundation of CP-1 (index idempotency) and CP-2 (symbol
id stability under cosmetic edits).

Design contract:

- Cosmetic edits — whitespace-only or comment-only changes — MUST produce the
  same ``content_hash``.
- Structural edits — rename, signature change, body change — MUST produce a
  different ``content_hash``.

Implementation strategy:

1. Remove single-line comments (``//``, ``#``).
2. Remove multi-line / block comments (``/* … */``, triple-quoted docstrings,
   Go doc comments).
3. Collapse all remaining whitespace sequences (spaces, tabs, newlines) to a
   single space.
4. Strip leading/trailing space.
5. SHA-256 the result; return the first 16 hex chars.

This approach is language-agnostic and intentionally simple.  It operates on
*raw text*, not on tree-sitter AST nodes, so it can also be used in tests
without a live tree-sitter installation.

The trade-off: over-zealous stripping of string literals that look like
comments is acceptable for the purpose of stable hashing — it errs on the side
of fewer spurious re-index events.
"""

from __future__ import annotations

import hashlib
import re

# ---------------------------------------------------------------------------
# Comment stripping patterns
# ---------------------------------------------------------------------------

# Single-line comments: // … (TypeScript/Go) and # … (Python)
_SINGLE_LINE_COMMENT = re.compile(r"(//|#)[^\n]*")

# Block comments: /* … */  (TypeScript/Go)
_BLOCK_COMMENT = re.compile(r"/\*.*?\*/", re.DOTALL)

# Python triple-quoted strings used as docstrings.
# We strip them when they appear as standalone expression statements (first
# non-whitespace on a line), which covers the docstring pattern.
# This is *not* a full Python parser; it handles the common case.
_TRIPLE_DOUBLE = re.compile(r'""".*?"""', re.DOTALL)
_TRIPLE_SINGLE = re.compile(r"'''.*?'''", re.DOTALL)

# Collapse whitespace
_WHITESPACE = re.compile(r"\s+")


def normalize_body(text: str) -> str:
    """Return a whitespace- and comment-stripped version of *text*.

    Suitable for hashing to detect structural (non-cosmetic) changes.
    """
    # Strip block comments first (they can span lines)
    text = _BLOCK_COMMENT.sub(" ", text)
    # Strip Python triple-quoted docstrings
    text = _TRIPLE_DOUBLE.sub(" ", text)
    text = _TRIPLE_SINGLE.sub(" ", text)
    # Strip single-line comments
    text = _SINGLE_LINE_COMMENT.sub(" ", text)
    # Collapse whitespace
    text = _WHITESPACE.sub(" ", text)
    return text.strip()


def content_hash(body_text: str) -> str:
    """Return ``sha256(normalize(body_text))[:16]``.

    This is the ``short_hash`` component of a symbol's stable ID per the design
    *ID conventions* section::

        symbol.id = "<lang>:<file_path>:<qualified_name>@<short_hash>"
        content_hash = sha256(normalized_body)[:16]
    """
    normalized = normalize_body(body_text)
    digest = hashlib.sha256(normalized.encode("utf-8")).hexdigest()
    return digest[:16]


def make_symbol_id(lang: str, file_path: str, qualified_name: str, body_text: str) -> str:
    """Construct the stable symbol ID per design conventions.

    Args:
        lang: Language prefix, e.g. ``"ts"``, ``"py"``, ``"go"``.
        file_path: Repo-relative path with forward slashes.
        qualified_name: Fully qualified name, e.g. ``"src/auth/jwt.validate"``.
        body_text: Raw body text of the symbol (before normalization).

    Returns:
        ``"<lang>:<file_path>:<qualified_name>@<short_hash>"``
    """
    short_hash = content_hash(body_text)
    return f"{lang}:{file_path}:{qualified_name}@{short_hash}"

File: crates/cognis-indexer/src/parser/normalize.rs
//! AST-text normalizer + content-hash / symbol-id utilities.
//!
//! Rust mirror of `packages/indexer/cognis_indexer/parsers/_normalize.py`. This
//! is the foundation of CP-1 (index idempotency) and CP-2 (symbol id stability
//! under cosmetic edits):
//!
//! - Cosmetic edits (whitespace-only / comment-only) MUST yield the same
//!   `content_hash`.
//! - Structural edits (rename / signature / body change) MUST yield a different
//!   `content_hash`.
//!
//! The normalizer operates on raw text (not tree-sitter nodes) and applies the
//! exact same passes, in the same order, as the Python implementation so the
//! Rust engine produces byte-identical symbol ids on the same input
//! (Requirement 9.2 parity).

use std::sync::OnceLock;

use regex::Regex;
use sha2::{Digest, Sha256};

/// `short_hash` length (hex chars) appended to a symbol id.
const SHORT_HASH_LEN: usize = 16;

// Compiled lazily once; each mirrors a pattern in `_normalize.py`.
fn block_comment() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    // `/* … */` (TypeScript/Go/C#/Java). `(?s)` = DOTALL so it spans lines.
    RE.get_or_init(|| Regex::new(r"(?s)/\*.*?\*/").unwrap())
}

fn triple_double() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r#"(?s)""".*?""""#).unwrap())
}

fn triple_single() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"(?s)'''.*?'''").unwrap())
}

fn single_line_comment() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    // `// …` (TS/Go/C#/Java) and `# …` (Python), up to end of line.
    RE.get_or_init(|| Regex::new(r"(//|#)[^\n]*").unwrap())
}

fn whitespace() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"\s+").unwrap())
}

/// Return a whitespace- and comment-stripped version of `text`.
///
/// Pass order matches `_normalize.py::normalize_body` exactly: block comments,
/// triple-double, triple-single, single-line, then whitespace collapse + trim.
pub fn normalize_body(text: &str) -> String {
    let text = block_comment().replace_all(text, " ");
    let text = triple_double().replace_all(&text, " ");
    let text = triple_single().replace_all(&text, " ");
    let text = single_line_comment().replace_all(&text, " ");
    let text = whitespace().replace_all(&text, " ");
    text.trim().to_string()
}

/// Return `sha256(normalize(body_text))[:16]` — the `short_hash` id component.
pub fn content_hash(body_text: &str) -> String {
    let normalized = normalize_body(body_text);
    let digest = Sha256::digest(normalized.as_bytes());
    let hex = hex_lower(&digest);
    hex[..SHORT_HASH_LEN].to_string()
}

/// Construct the stable symbol id: `<lang>:<file_path>:<qualified_name>@<short_hash>`.
pub fn make_symbol_id(
    lang: &str,
    file_path: &str,
    qualified_name: &str,
    body_text: &str,
) -> String {
    format!(
        "{lang}:{file_path}:{qualified_name}@{}",
        content_hash(body_text)
    )
}

fn hex_lower(bytes: &[u8]) -> String {
    use std::fmt::Write as _;
    let mut s = String::with_capacity(bytes.len() * 2);
    for b in bytes {
        let _ = write!(s, "{b:02x}");
    }
    s
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn whitespace_only_edit_is_stable() {
        let a = "def foo():\n    return 1";
        let b = "def   foo():\n\n        return 1\n";
        assert_eq!(content_hash(a), content_hash(b));
    }

    #[test]
    fn comment_only_edit_is_stable() {
        let a = "fn foo() { return 1; } // old comment";
        let b = "fn foo() { return 1; } // a totally different comment";
        assert_eq!(content_hash(a), content_hash(b));
    }

    #[test]
    fn block_and_docstring_comments_stripped() {
        let a = "/* header */ fn foo() {}";
        let b = "fn foo() {}";
        assert_eq!(content_hash(a), content_hash(b));
        let c = "def f():\n    \"\"\"doc one\"\"\"\n    return 1";
        let d = "def f():\n    \"\"\"doc two changed\"\"\"\n    return 1";
        assert_eq!(content_hash(c), content_hash(d));
    }

    #[test]
    fn structural_edit_changes_hash() {
        assert_ne!(content_hash("fn foo() {}"), content_hash("fn bar() {}"));
    }

    #[test]
    fn symbol_id_shape() {
        let id = make_symbol_id("py", "src/m.py", "m.foo", "def foo(): pass");
        assert!(id.starts_with("py:src/m.py:m.foo@"));
        let hash = id.rsplit('@').next().unwrap();
        assert_eq!(hash.len(), SHORT_HASH_LEN);
        assert!(hash.chars().all(|c| c.is_ascii_hexdigit()));
    }
}
